{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python385jvsc74a57bd07b9569140d99d56f18a75135eb8a624439e86e9b615bd911525c395d062533e1",
   "display_name": "Python 3.8.5 64-bit ('base': conda)"
  },
  "metadata": {
   "interpreter": {
    "hash": "7b9569140d99d56f18a75135eb8a624439e86e9b615bd911525c395d062533e1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "# What we need:\n",
    "1) lib folder for container libraries<br>\n",
    "2) file handling mechanism<br>\n",
    "3) Exception handling<br>"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Event, outputs/Document.txt has been created\n",
      "Event, outputs/Document.txt has been deleted\n",
      "Event, outputs/output_genome1 has been deleted\n",
      "Event, outputs/output_genome2 has been deleted\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "from watchdog.observers import Observer\n",
    "from watchdog.events import PatternMatchingEventHandler\n",
    "\n",
    "# Event handler\n",
    "patterns = \"*\"\n",
    "ignore_patterns = \"\"\n",
    "ignore_directories = False\n",
    "case_sensitive = False\n",
    "automation_event_handler = PatternMatchingEventHandler(patterns,ignore_patterns,ignore_directories,case_sensitive)\n",
    "\n",
    "def on_created(event):\n",
    "    print(f\"Event, {event.src_path} has been created\")\n",
    "\n",
    "def on_deleted(event):\n",
    "    print(f\"Event, {event.src_path} has been deleted\")\n",
    "\n",
    "automation_event_handler.on_created = on_created\n",
    "automation_event_handler.on_deleted = on_deleted\n",
    "\n",
    "path=\"outputs\"\n",
    "go_recursively = True\n",
    "automation_observer = Observer()\n",
    "automation_observer.schedule(automation_event_handler, path, recursive=go_recursively)\n",
    "automation_observer.start()\n",
    "try:\n",
    "    while True:\n",
    "        time.sleep(1)\n",
    "except KeyboardInterrupt:\n",
    "    automation_observer.stop()\n",
    "    automation_observer.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'logs/automatic_mode_execution.txt'",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-f4c36db21946>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# Appending to log file for execution status.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0mlog\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"logs/automatic_mode_execution.txt\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"a\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\nExecution start time: [ {} ]\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'logs/automatic_mode_execution.txt'"
     ]
    }
   ],
   "source": [
    "import time\n",
    "import os\n",
    "import os.path\n",
    "from watchdog.observers import Observer\n",
    "from watchdog.events import PatternMatchingEventHandler\n",
    "from lib.hisat2_docker import hisat2\n",
    "from lib.stringtie_docker import stringtie\n",
    "from lib.deseq2_docker import deseq2\n",
    "from lib.bioconda_py import bioconda\n",
    "from lib.fastx_toolkit_docker import fastx_toolkit\n",
    "\n",
    "input_command = None\n",
    "\n",
    "# Appending to log file for execution status.\n",
    "log = open(\"logs/automatic_mode_execution.txt\",\"a\")\n",
    "log.write(\"\\nExecution start time: [ {} ]\".format(str(time.time())))\n",
    "\n",
    "# Declaring counters for execution tracking\n",
    "cnt = 0\n",
    "cnt_bioconda = 0\n",
    "cnt_hisat2 = 0\n",
    "cnt_stringtie = 0\n",
    "cnt_deseq2 = 0\n",
    "cnt_fastx = 0\n",
    "\n",
    "# --------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Changing path to Output directory\n",
    "path = os.path.abspath(os.path.join(os.getcwd(),'outputs'))\n",
    "os.chdir(path)\n",
    "\n",
    "# Reading counter number previously appended to file\n",
    "f = open(\"counter.txt\",\"r\")\n",
    "a = f.read()\n",
    "b = int(a)+1\n",
    "\n",
    "# Writing a new number to be appended to current file.\n",
    "f1 = open(\"counter.txt\",\"w\")\n",
    "f1.write(str(b))\n",
    "f1.close()\n",
    "\n",
    "f_name = \"output_genome\" + str(b) # new appended folder name\n",
    "os.mkdir(f_name) # create folder\n",
    "log.write(\"\\nOutput folder name: {}\".format(f_name))\n",
    "os.chdir(f_name) # Change program execution folder to newly created folder\n",
    "\n",
    "# ----------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Running the containers\n",
    "def printing_return(fun):\n",
    "    input_command = input('Enter the Container Command: ')\n",
    "    log.write('\\nContainer command used: {}'.format(input_command))\n",
    "\n",
    "    if fun == 'hisat2':\n",
    "        output_hisat = hisat2(input_command)\n",
    "        cnt_hisat2 += 1\n",
    "        print(output_hisat)\n",
    "    if fun == 'stringtie':\n",
    "        output_stringtie = stringtie(input_command)\n",
    "        cnt_stringtie += 1\n",
    "        print(output_stringtie)\n",
    "    if fun == 'deseq2':\n",
    "        output_deseq2 = deseq2(input_command)\n",
    "        cnt_deseq2 += 1\n",
    "        print(output_deseq2)\n",
    "    if fun == 'fastx_toolkit':\n",
    "        output_fastx = fastx_toolkit(input_command)\n",
    "        cnt_fastx += 1\n",
    "        print(output_fastx)\n",
    "    if fun == 'bioconda':\n",
    "        output_bioconda = bioconda(input_command)\n",
    "        cnt_bioconda += 1\n",
    "        print(output_bioconda)\n",
    "\n",
    "# ---------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "printing_return(hisat2)\n",
    "\n",
    "# ----------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Event handler and automation of pipeline\n",
    "patterns = \"*\"\n",
    "ignore_patterns = \"\"\n",
    "ignore_directories = False\n",
    "case_sensitive = False\n",
    "automation_event_handler = PatternMatchingEventHandler(patterns,ignore_patterns,ignore_directories,case_sensitive)\n",
    "\n",
    "def on_created(event):\n",
    "    #print(f\"Event, {event.src_path} has been created\")\n",
    "    if event.src_path.endswith('.ht2'):\n",
    "        printing_return(hisat2)\n",
    "    if event.src_path.endswith('.sam') or event.src_path.endswith('.bam'):\n",
    "        printing_return(stringtie)\n",
    "    if event.src_path.endswith('.gtf'):\n",
    "        printing_return(deseq2)\n",
    "    if event.src_path.endswith('.csv'):\n",
    "        print(\"Pipeline generated csv succesfully\")\n",
    "        automation_observer.stop()\n",
    "        automation_observer.join()\n",
    "\n",
    "\n",
    "def on_deleted(event):\n",
    "    print(f\"Event, {event.src_path} has been deleted\")\n",
    "\n",
    "automation_event_handler.on_created = on_created\n",
    "automation_event_handler.on_deleted = on_deleted\n",
    "\n",
    "path=\".\"\n",
    "go_recursively = True\n",
    "automation_observer = Observer()\n",
    "automation_observer.schedule(automation_event_handler, path, recursive=go_recursively)\n",
    "automation_observer.start()\n",
    "try:\n",
    "    while True:\n",
    "        time.sleep(1)\n",
    "except KeyboardInterrupt:\n",
    "    automation_observer.stop()\n",
    "    automation_observer.join()\n",
    "\n",
    "# --------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# Writing logs\n",
    "log.write('\\nBioconda executed {} times'.format(cnt_bioconda))\n",
    "log.write('\\nhisat2 executed {} times'.format(cnt_hisat2))\n",
    "log.write('\\nstringtie executed {} times'.format(cnt_stringtie))\n",
    "log.write('\\ndeseq2 executed {} times'.format(cnt_deseq2))\n",
    "log.write('\\nfastx_toolkit executed {} times'.format(cnt_fastx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}