{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python385jvsc74a57bd07b9569140d99d56f18a75135eb8a624439e86e9b615bd911525c395d062533e1",
   "display_name": "Python 3.8.5 64-bit ('base': conda)"
  },
  "metadata": {
   "interpreter": {
    "hash": "7b9569140d99d56f18a75135eb8a624439e86e9b615bd911525c395d062533e1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "# What we need:\n",
    "1) lib folder for container libraries<br>\n",
    "2) file handling mechanism<br>\n",
    "3) Exception handling<br>"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "from watchdog.observers import Observer\n",
    "from watchdog.events import PatternMatchingEventHandler\n",
    "\n",
    "# Event handler\n",
    "patterns = \"*\"\n",
    "ignore_patterns = \"\"\n",
    "ignore_directories = False\n",
    "case_sensitive = False\n",
    "automation_event_handler = PatternMatchingEventHandler(patterns,ignore_patterns,ignore_directories,case_sensitive)\n",
    "\n",
    "def on_created(event):\n",
    "    print(f\"Event, {event.src_path} has been created\")\n",
    "\n",
    "def on_deleted(event):\n",
    "    print(f\"Event, {event.src_path} has been deleted\")\n",
    "\n",
    "automation_event_handler.on_created = on_created\n",
    "automation_event_handler.on_deleted = on_deleted\n",
    "\n",
    "path=\"outputs\"\n",
    "go_recursively = True\n",
    "automation_observer = Observer()\n",
    "automation_observer.schedule(automation_event_handler, path, recursive=go_recursively)\n",
    "automation_observer.start()\n",
    "try:\n",
    "    while True:\n",
    "        time.sleep(1)\n",
    "except KeyboardInterrupt:\n",
    "    automation_observer.stop()\n",
    "    automation_observer.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import os\n",
    "import os.path\n",
    "from watchdog.observers import Observer\n",
    "from watchdog.events import PatternMatchingEventHandler\n",
    "\n",
    "from lib.hisat2_docker import hisat2\n",
    "from lib.stringtie_docker import stringtie\n",
    "from lib.deseq2_docker import deseq2\n",
    "from lib.bioconda_py import bioconda\n",
    "from lib.fastx_toolkit_docker import fastx_toolkit\n",
    "\n",
    "class auto_exec:\n",
    "    input_command = None\n",
    "\n",
    "    # Appending to log file for execution status.\n",
    "    log = open(\"logs/automatic_mode_execution.txt\",\"a\")\n",
    "    log.write(\"\\nExecution start time: [ {} ]\".format(str(time.time())))\n",
    "\n",
    "    # Declaring counters for execution tracking\n",
    "    cnt = 0\n",
    "    cnt_bioconda = 0\n",
    "    cnt_hisat2 = 0\n",
    "    cnt_stringtie = 0\n",
    "    cnt_deseq2 = 0\n",
    "    cnt_fastx = 0\n",
    "\n",
    "    multithread_cnt = 0\n",
    "\n",
    "    # --------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "    # Changing path to Output directory\n",
    "    path = os.path.abspath(os.path.join(os.getcwd(),'outputs'))\n",
    "    os.chdir(path)\n",
    "\n",
    "    # Reading counter number previously appended to file\n",
    "    f = open(\"counter.txt\",\"r\")\n",
    "    a = f.read()\n",
    "    b = int(a)+1\n",
    "\n",
    "    # Writing a new number to be appended to current file.\n",
    "    f1 = open(\"counter.txt\",\"w\")\n",
    "    f1.write(str(b))\n",
    "    f1.close()\n",
    "\n",
    "    f_name = \"output_genome\" + str(b) # new appended folder name\n",
    "    os.mkdir(f_name) # create folder\n",
    "    log.write(\"\\nOutput folder name: {}\".format(f_name))\n",
    "    os.chdir(f_name) # Change program execution folder to newly created folder\n",
    "\n",
    "    # ----------------------------------------------------------------------------------------------------------------------\n",
    "    def __init__(self):\n",
    "        self.printing_return('hisat2')\n",
    "        self.event_handling()\n",
    "        self.write_log()\n",
    "\n",
    "    # ----------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "    # Running the containers\n",
    "    def printing_return(self, fun):\n",
    "        self.input_command = input('Enter the Container Command: ')\n",
    "        self.log.write('\\nContainer command used: {}'.format(self.input_command))\n",
    "\n",
    "        if fun == 'hisat2':\n",
    "            output_hisat = hisat2(self.input_command)\n",
    "            self.log.write(\"hisat2 command used: {}\".format(self.input_command))\n",
    "            self.cnt_hisat2 += 1\n",
    "            print(output_hisat)\n",
    "        if fun == 'stringtie':\n",
    "            output_stringtie = stringtie(self.input_command)\n",
    "            self.log.write(\"stringtie command used: {}\".format(self.input_command))\n",
    "            self.cnt_stringtie += 1\n",
    "            print(output_stringtie)\n",
    "        if fun == 'deseq2':\n",
    "            output_deseq2 = deseq2(self.input_command)\n",
    "            self.log.write(\"deseq2 command used: {}\".format(self.input_command))\n",
    "            self.cnt_deseq2 += 1\n",
    "            print(output_deseq2)\n",
    "        if fun == 'fastx_toolkit':\n",
    "            output_fastx = fastx_toolkit(self.input_command)\n",
    "            self.log.write(\"fast_toolkit command used: {}\".format(self.input_command))\n",
    "            self.cnt_fastx += 1\n",
    "            print(output_fastx)\n",
    "        if fun == 'bioconda':\n",
    "            output_bioconda = bioconda(self.input_command)\n",
    "            self.log.write(\"bioconda command used: {}\".format(self.input_command))\n",
    "            self.cnt_bioconda += 1\n",
    "            print(output_bioconda)\n",
    "\n",
    "    # ---------------------------------------------------------------------------------------------------------------------------s\n",
    "\n",
    "    # Event handler and automation of pipeline\n",
    "    def event_handling(self):\n",
    "        patterns = \"*\"\n",
    "        ignore_patterns = \"\"\n",
    "        ignore_directories = False\n",
    "        case_sensitive = False\n",
    "        automation_event_handler = PatternMatchingEventHandler(patterns,ignore_patterns,ignore_directories,case_sensitive)\n",
    "\n",
    "        def on_created(event):\n",
    "            #print(f\"Event, {event.src_path} has been created\")\n",
    "            if event.src_path.endswith('.ht2'):\n",
    "                self.printing_return('hisat2')\n",
    "            if event.src_path.endswith('.sam') or event.src_path.endswith('.bam'):\n",
    "                self.printing_return('stringtie')\n",
    "            if event.src_path.endswith('.gtf'):\n",
    "                self.printing_return('deseq2')\n",
    "            if event.src_path.endswith('.csv'):\n",
    "                print(\"Pipeline generated csv succesfully\")\n",
    "                self.multithread_cnt = 1\n",
    "\n",
    "\n",
    "        def on_deleted(event):\n",
    "            print(f\"Event, {event.src_path} has been deleted\")\n",
    "\n",
    "        automation_event_handler.on_created = on_created\n",
    "        automation_event_handler.on_deleted = on_deleted\n",
    "\n",
    "        path=\".\"\n",
    "        go_recursively = True\n",
    "        automation_observer = Observer()\n",
    "        automation_observer.schedule(automation_event_handler, path, recursive=go_recursively)\n",
    "        automation_observer.start()\n",
    "\n",
    "        try:\n",
    "            while True:\n",
    "                if self.multithread_cnt == 0:\n",
    "                    time.sleep(1)\n",
    "                else:\n",
    "                    exit()\n",
    "\n",
    "        except KeyboardInterrupt:\n",
    "            automation_observer.stop()\n",
    "            automation_observer.join()\n",
    "\n",
    "    # --------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "    # Writing logs\n",
    "    def write_log(self):\n",
    "        self.log.write('\\nBioconda executed {} times'.format(self.cnt_bioconda))\n",
    "        self.log.write('\\nhisat2 executed {} times'.format(self.cnt_hisat2))\n",
    "        self.log.write('\\nstringtie executed {} times'.format(self.cnt_stringtie))\n",
    "        self.log.write('\\ndeseq2 executed {} times'.format(self.cnt_deseq2))\n",
    "        self.log.write('\\nfastx_toolkit executed {} times'.format(self.cnt_fastx)) \n",
    "\n",
    "class test:\n",
    "    a = auto_exec()"
   ]
  }
 ]
}